File: programmers/ForKaKAO/2021/level3/합승 택시 요금.cpp
#include <string>
#include <vector>

using namespace std;

#define INF 50000000

int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int answer = INF;
    int dist[201][201];
    
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= n; j++) {
            if (i == j) {
                dist[i][j] = 0;
            } else {
                dist[i][j] = INF;
            }
        }
    }
    for (auto f : fares) {
        dist[f[0]][f[1]] = f[2];
        dist[f[1]][f[0]] = f[2];
    }
    
    // 플로이드 와샬
    for (int k = 1; k <= n; k++) {
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (dist[i][j] > dist[i][k] + dist[k][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }
    
    for (int i = 1; i <= n; i++) {
        answer = min(answer, dist[s][i] + dist[i][a] + dist[i][b]);
    }
    
    return answer;
}

File: programmers/ForKaKAO/2021/level3/합승 택시 요금.py
def solution(n, s, a, b, fares):
    INF = 987654321
    answer = INF
    dist = [[INF for _ in range(n)] for _ in range(n)]
    for i in range(n):
        dist[i][i] = 0
    
    for e in fares:
        dist[e[0] - 1][e[1] - 1] = e[2]
        dist[e[1] - 1][e[0] - 1] = e[2]
        
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if dist[i][j] > dist[i][k] + dist[k][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]
    
    s -= 1
    a -= 1
    b -= 1
    
    for k in range(n):
        answer = min(answer, dist[s][k] + dist[k][b] + dist[k][a])
        
    return answer